v1/handler.js

```diff
@@ -0,0 +1,203 @@
+        return;
+    }
+    var remote = String(resFrame.arg2);
+    self.remoteName = remote;
+    self.emit('ident.out', remote);
+};
+
+TChannelHandler.prototype.handleCallRequest = function handleCallRequest(reqFrame) {
+    var self = this;
+    var id = reqFrame.header.id;
+    var name = reqFrame.arg1.toString();
+
+    if (name === 'TChannel identify') {
+        self.handleInitRequest(reqFrame);
+        return;
+    }
+
+    if (self.remoteName === null) {
+        self.emit('error', new Error('call request before init request')); // TODO typed error
+        return;
+    }
+
+    var handler = self.channel.getEndpointHandler(name);
+    self.runInOp(handler, {
+        id: id,
+        arg1: reqFrame.arg1,
+        arg2: reqFrame.arg2,
+        arg3: reqFrame.arg3,
+    }, function sendResponseFrame(err, res1, res2, callback) {
+        self.sendResponseFrame(reqFrame, err, res1, res2, callback);
+    });
+};
+
+TChannelHandler.prototype.handleCallResponse = function handleCallResponse(resFrame) {
+    var self = this;
+
+    if (String(resFrame.arg1) === 'TChannel identify') {
+        self.handleInitResponse(resFrame);
+        return;
+    }
+
+    if (self.remoteName === null) {
+        self.emit('error', new Error('call response before init response')); // TODO typed error
+        return;
+    }
+
+    var id = resFrame.header.id;
+    var arg2 = resFrame.arg2;
+    var arg3 = resFrame.arg3;
+    self.completeOutOp(null, id, arg2, arg3);
+};
+
+TChannelHandler.prototype.handleError = function handleError(errFrame) {
+    var self = this;
+    var id = errFrame.header.id;
+    var message = errFrame.arg1;
+    var err = new Error(message);
+    self.completeOutOp(err, id, null, null);
+};
+
+TChannelHandler.prototype.sendInitRequest = function sendInitRequest() {
+    var self = this;
+    var reqFrame = new v1.Frame();
+    var id = self.nextFrameId();
+    reqFrame.header.id = id;
+    reqFrame.header.seq = 0;
+    reqFrame.set('TChannel identify', self.channel.hostPort, null);
+    reqFrame.header.type = v1.Types.reqCompleteMessage;
+    self.writeFrame(reqFrame);
+};
+
+TChannelHandler.prototype.sendInitResponse = function sendInitResponse(reqFrame) {
+    var self = this;
+    var id = reqFrame.header.id;
+    var arg1 = reqFrame.arg1;
+    var resFrame = new v1.Frame();
+    resFrame.header.id = id;
+    resFrame.header.seq = 0;
+    resFrame.set(arg1, self.channel.hostPort, null);
+    resFrame.header.type = v1.Types.resCompleteMessage;
+    self.writeFrame(resFrame);
+};
+
+TChannelHandler.prototype.sendRequestFrame = function sendRequestFrame(options, callback) {
+    var self = this;
+    var id = self.nextFrameId();
+    var reqFrame = new v1.Frame();
+    reqFrame.header.id = id;
+    reqFrame.header.seq = 0;
+    reqFrame.set(options.arg1, options.arg2, options.arg3);
+    reqFrame.header.type = v1.Types.reqCompleteMessage;
+    self.writeFrame(reqFrame, callback);
+    return id;
```

Comment:
In terms of interface design we may not want to return the id.

A common thing we want to do as well as pass the `callback` to `socket.write()` is to return the boolean result of `socket.write()` which is the "should I write more" backpressure boolean.

If we want the `TChannelHandler` interface to expose more information about the stream state we might want to return "should I write more" backpressure information in some fashion.